test/core/event_engine/posix/posix_endpoint_test.cc

```diff
@@ -0,0 +1,352 @@
+                         int num_connections,
+                         std::shared_ptr<EventEngine> posix_ee) {
+  EXPECT_NE(GetOracleEE(), nullptr);
+  EXPECT_NE(poller, nullptr);
+  std::list<std::tuple<std::unique_ptr<EventEngine::Endpoint>,
+                       std::unique_ptr<EventEngine::Endpoint>>>
+      connections;
+  auto memory_quota = absl::make_unique<grpc_core::MemoryQuota>("bar");
+  std::string target_addr = absl::StrCat(
+      "ipv6:[::1]:", std::to_string(grpc_pick_unused_port_or_die()));
+  EventEngine::ResolvedAddress resolved_addr =
+      URIToResolvedAddress(target_addr);
+  std::unique_ptr<EventEngine::Endpoint> server_endpoint;
+  grpc_core::Notification* server_signal = new grpc_core::Notification();
+
+  Listener::AcceptCallback accept_cb =
+      [&server_endpoint, &server_signal](
+          std::unique_ptr<Endpoint> ep,
+          grpc_core::MemoryAllocator /*memory_allocator*/) {
+        server_endpoint = std::move(ep);
+        server_signal->Notify();
+      };
+  grpc_core::ChannelArgs args;
+  auto quota = grpc_core::ResourceQuota::Default();
+  args = args.Set(GRPC_ARG_RESOURCE_QUOTA, quota);
+  if (is_zero_copy_enabled) {
+    args = args.Set(GRPC_ARG_TCP_TX_ZEROCOPY_ENABLED, 1);
+    args = args.Set(GRPC_ARG_TCP_TX_ZEROCOPY_SEND_BYTES_THRESHOLD,
+                    kMinMessageSize);
+  }
+  ChannelArgsEndpointConfig config(args);
+  auto listener = GetOracleEE()->CreateListener(
+      std::move(accept_cb),
+      [](absl::Status status) { ASSERT_TRUE(status.ok()); }, config,
+      absl::make_unique<grpc_core::MemoryQuota>("foo"));
+  GPR_ASSERT(listener.ok());
+
+  EXPECT_TRUE((*listener)->Bind(resolved_addr).ok());
+  EXPECT_TRUE((*listener)->Start().ok());
+
+  // Create client socket and connect to the target address.
+  for (int i = 0; i < num_connections; ++i) {
+    int client_fd = ConnectToServerOrDie(resolved_addr);
+    EventHandle* handle =
+        poller->CreateHandle(client_fd, "test", poller->CanTrackErrors());
+    EXPECT_NE(handle, nullptr);
+    server_signal->WaitForNotification();
+    EXPECT_NE(server_endpoint, nullptr);
+    ++g_num_active_connections;
+    connections.push_back(
+        std::make_tuple<std::unique_ptr<Endpoint>, std::unique_ptr<Endpoint>>(
+            CreatePosixEndpoint(handle,
+                                PosixEngineClosure::TestOnlyToClosure(
+                                    [poller](absl::Status /*status*/) {
+                                      if (--g_num_active_connections == 0) {
+                                        poller->Kick();
+                                      }
+                                    }),
+                                posix_ee, config),
+            std::move(server_endpoint)));
+    delete server_signal;
+    server_signal = new grpc_core::Notification();
+  }
+  delete server_signal;
+  return connections;
+}
+
+}  // namespace
+
+class TestParam {
```

Comment:
This looks like a struct to me, it has no owned behaviors, does not mutate its input, and exposes all of its state.

```
struct TestParam {
  std::string poller;
  bool is_zero_copy_enabled;
};
```

constructor optional.